### PR TITLE
feat: extract structured data from decision PDFs

### PR DESCRIPTION
**Pull request by kouloumos**
Greek municipal decision PDFs from Diavgeia contain valuable structured data — attendance lists, vote results, decision excerpts, legal references — that we currently don't capture. This PR adds a task that extracts all of it using Claude, with name matching against our people records.

It also handles a real-world complication: members arrive late or leave early, and subjects are sometimes discussed out of agenda order. The extracted attendance changes are combined with discussion order to compute who was actually present for each subject.

## Test plan
-  `npm test` && `npm run typecheck`
- CLI: `npm run cli extract-decision <pdf-url>` with various municipalities
- End-to-end via opencouncil admin panel

---

> [!NOTE]
> **Medium Risk**
> Adds a sizable new LLM-driven PDF extraction pipeline and extends an existing task’s API contract/results, increasing risk of correctness/regression and cost/latency changes during polling.
> 
> **Overview**
> `pollDecisions` now goes beyond matching decisions: it optionally **extracts structured data from matched decision PDFs** (excerpt, references, attendance, vote results/details, mayor presence) and returns it in a new `extractions` payload alongside aggregated LLM token `costs`.
> 
> This adds a new PDF/LLM extraction subsystem (`decisionPdfExtraction`, `extractionPipeline`, `effectiveAttendance`, `decisionValidation`) including an on-disk cache, progressive page extraction for large PDFs, two-step people name matching (token-sort then Haiku fallback), vote inference, and verification/reassignment when PDF agenda numbering contradicts the match.
> 
> The CLI gains `extract-decision` for ad-hoc extraction (ADA/URL/local file) with warnings and effective-attendance computation, while `aiChat` now supports configurable `maxTokens`, centralized `HAIKU_MODEL`, and a `formatUsage` helper. Docs are cleaned up (remove `docs/pollDecisions.md`, update references), and `pdf-lib` is added as a dependency (with updated lockfile/Nix hash).
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit a79a6e63dceaed0098cc900d71c8bc980bbf28bd. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

<h3>Greptile Summary</h3>

This PR adds a comprehensive decision PDF extraction pipeline: it downloads Greek municipal decision PDFs, uses Claude to parse structured data (attendance, votes, references, excerpts), matches names to known `people` records via token-sort + LLM fallback, computes effective per-subject attendance accounting for arrivals/departures and agenda reordering, and infers implicit FOR votes. It integrates into `pollDecisions` with a two-pass verification/reassignment step, adds a `needsExtraction` re-run path, and extends the CLI with an `extract-decision` command.

Previous review concerns (HTTP error checking, duplicate `unmatchedMembers`, LLM JSON name-field validation, off-by-one tail extraction, `needsExtraction` verification bypass) are all fixed in the latest commit.

<h3>Confidence Score: 5/5</h3>

Safe to merge — all previously identified P0/P1 issues are resolved, remaining findings are P2 style/UX improvements.

All critical issues from the previous review round are fixed: HTTP response check, duplicate unmatchedMembers, LLM JSON validation, off-by-one tail extraction, and needsExtraction verification bypass. The two new comments are both P2 — one about caching incomplete results forever (forceExtract is the documented escape hatch), and one about warning clarity when opposition voters are unmatched. The extraction logic, name matching, effective attendance computation, and vote inference are all thoroughly unit-tested with no observed correctness gaps.

src/tasks/utils/decisionPdfExtraction.ts (incomplete result caching), src/tasks/utils/extractionPipeline.ts (NO_VOTE_DETAILS misleading message when voters are unmatched)

<h3>Important Files Changed</h3>

| Filename | Overview |
|----------|----------|
| src/tasks/utils/decisionPdfExtraction.ts | Core PDF download, progressive page extraction, Claude API calls, and name matching (token-sort + LLM fallback); HTTP error check, LLM JSON validation, and tail off-by-one are all fixed. Cache writes incomplete results permanently. |
| src/tasks/utils/extractionPipeline.ts | Orchestrates batch PDF extraction, meeting-level name matching, vote inference, and decision result assembly; unmatched dedup and per-decision warnings are correct. |
| src/tasks/utils/effectiveAttendance.ts | Computes per-subject effective attendance by walking the discussion sequence and applying arrival/departure changes; logic is correct and well-tested. |
| src/tasks/utils/decisionValidation.ts | Validates raw Claude extraction and post-inference vote data; warning codes and severities are well-defined and correctly assigned. |
| src/tasks/pollDecisions.ts | Extended with PDF extraction phase, two-pass mismatch verification, needsExtraction re-run path, and Diavgeia metadata enrichment; re-extraction subjects are correctly exempted from the verification loop. |
| src/types.ts | ExtractedDecisionResult extended with subjectInfo, fromCache, warnings, protocolNumber, and Diavgeia metadata fields; decisionNumber is now surfaced as protocolNumber. |
| src/tasks/utils/decisionPdfExtraction.test.ts | Comprehensive unit tests for name matching, vote inference, and progressive PDF extraction; covers edge cases including Greek diacritics, nickname variants, and multi-step matching. |
| src/tasks/utils/effectiveAttendance.test.ts | Well-structured tests covering all attendance change scenarios including reordering, during/after timing, and out-of-agenda items. |
| src/tasks/utils/extractionPipeline.test.ts | Pipeline integration tests covering warning propagation, vote deduplication, and LLM fallback behavior; intentionally documents full people list being passed to LLM. |

</details>

<h3>Sequence Diagram</h3>

```mermaid
sequenceDiagram
    participant Caller
    participant pollDecisions
    participant extractionPipeline
    participant decisionPdfExtraction
    participant Claude
    participant cache as /tmp cache

    Caller->>pollDecisions: PollDecisionsRequest (people, subjects, forceExtract?)
    pollDecisions->>pollDecisions: Phase 1 - Diavgeia title/ADA matching
    pollDecisions->>extractionPipeline: extractDecisionsFromPdfs(subjects, people)

    loop Batched PDF extraction (BATCH_SIZE=5)
        extractionPipeline->>decisionPdfExtraction: extractDecisionFromPdf(pdfUrl, mayorName, skipCache)
        decisionPdfExtraction->>cache: readCache(pdfUrl)
        alt cache hit
            cache-->>decisionPdfExtraction: RawExtractedDecision
        else cache miss / small PDF
            decisionPdfExtraction->>Claude: full PDF (base64)
            Claude-->>decisionPdfExtraction: RawExtractedDecision JSON
            decisionPdfExtraction->>cache: writeCache(result)
        else cache miss / large PDF
            loop progressive: 5 to 15 front pages
                decisionPdfExtraction->>Claude: partial PDF pages
                Claude-->>decisionPdfExtraction: result (incomplete?)
            end
            opt tail fallback (tailStart >= MAX_FRONT_PAGES)
                decisionPdfExtraction->>Claude: last 5 pages
                Claude-->>decisionPdfExtraction: tail result merged with front
            end
            decisionPdfExtraction->>cache: writeCache(result)
        end
        decisionPdfExtraction-->>extractionPipeline: RawExtractedDecision + usage + fromCache
    end

    extractionPipeline->>extractionPipeline: Phase 2 - Meeting-level name matching
    note over extractionPipeline: Step 1 token-sort normalized Greek names
    note over extractionPipeline: Step 2 LLM fallback via Haiku full people list

    loop Per-PDF Phase 3 - Build results
        extractionPipeline->>extractionPipeline: processRawExtraction effective attendance + vote inference
        extractionPipeline->>extractionPipeline: validateRawExtraction + validateProcessedDecision
    end

    extractionPipeline-->>pollDecisions: ExtractionPipelineResult decisions warnings usage
    pollDecisions->>pollDecisions: Two-pass verification skip needsExtraction subjects
    pollDecisions->>pollDecisions: Enrich with Diavgeia metadata
    pollDecisions-->>Caller: PollDecisionsResult matches extractions costs
```

<details><summary>Prompt To Fix All With AI</summary>

`````markdown
This is a comment left during a code review.
Path: src/tasks/utils/decisionPdfExtraction.ts
Line: 583-587

Comment:
**Incomplete results cached indefinitely**

When progressive extraction exhausts both front and tail pages without finding the decision section, the incomplete result is written to cache. Any subsequent run for the same PDF URL without `forceExtract: true` will silently return the cached incomplete result — even if the transient issue that caused truncation (e.g. a flaky Diavgeia server) is now resolved. The `EXTRACTION_INCOMPLETE` warning and `fromCache: true` are present, but operators may not intuitively know to use `forceExtract` to break out of this state.

Consider skipping the cache write when `incomplete: true`:

```typescript
const bestResult: RawExtractedDecision = { ...lastFrontResult!, incomplete: true };
// Don't cache permanently incomplete results — they should be retried on the next run
// writeCache(pdfUrl, bestResult);
return { result: bestResult, usage: totalUsage, fromCache: false };
```

How can I resolve this? If you propose a fix, please make it concise.

---

This is a comment left during a code review.
Path: src/tasks/utils/extractionPipeline.ts
Line: 196-198

Comment:
**`NO_VOTE_DETAILS` fires when unmatched voters held opposition votes**

When a majority-vote decision has opposition voters that can't be matched to any person (they end up in `unmatchedMembers`), their votes are simply dropped from `voteDetails` (line 197). `validateProcessedDecision` then sees a `voteDetails` array with only inferred FOR entries, triggering the `NO_VOTE_DETAILS` warning. The warning message — "no AGAINST or ABSTAIN voters were found" — is technically correct but misleading: AGAINST/ABSTAIN voters _were_ found in the PDF, they just couldn't be matched.

Callers inspecting both `warnings` and `unmatchedMembers` can piece this together, but combining the warning with a note like "N unmatched voters with opposition votes" would make the signal unambiguous.

How can I resolve this? If you propose a fix, please make it concise.
`````

</details>

<sub>Reviews (19): Last reviewed commit: ["fix: address bot review comments on extr..."](https://github.com/schemalabz/opencouncil-tasks/commit/a79a6e63dceaed0098cc900d71c8bc980bbf28bd) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=25854423)</sub>

### CODE REVIEW

**Comment by kouloumos** [comment]
Main has moved forward (38 commits from the summarize-v2 merge) and the branch was out of sync. I've rebased all 16 commits on top of the current main and force-pushed. Conflicts were in `summarize.ts` (old code dropped by the v2 rewrite), `ai.ts` (merged the new `maxTokens` param with main's `tools`/`outputFormat`/`cacheSystemPrompt`), and `types.ts` (trivial).

The original branch state is preserved at [`backup/pr-33-original`](https://github.com/kouloumos/opencouncil-tasks/tree/backup/pr-33-original).

**Comment by kouloumos** [comment]
Post-rebase review found a few bugs in the extraction logic — fixed in the latest push:

**1. Duplicate vote entries from name casing mismatch** (`inferForVotes`)
Attendance lists use uppercase (`ΠΑΠΑΔΟΠΟΥΛΟΣ ΙΩΑΝΝΗΣ`) while vote sections use mixed case (`Παπαδόπουλος Ιωάννης`). Exact string matching caused members with explicit AGAINST votes to also get an inferred FOR vote. Fixed by normalizing names (strip accents, lowercase) before comparing.

**2. Tail-page merge losing `discussionOrder` and `subjectInfo`**
For large PDFs (>15 pages) where the decision section is in the last pages, the merge used the tail result as the spread base, overwriting `discussionOrder` and `subjectInfo` extracted from the front pages with `null`. This broke effective attendance computation when discussion order was reordered.

**3. Verification incorrectly removing re-extraction subjects**
Subjects with `needsExtraction` (linked decision but failed PDF extraction on a previous run) were not in the `matches` array. The verification loop processed them anyway, and on a `subjectInfo` mismatch, `removeMatchAndExtraction` discarded the freshly extracted data and marked the subject as unmatched.

Also centralized model IDs as named constants (`EXTRACTION_MODEL`, `HAIKU_MODEL`) and forwarded `maxTokens` in `aiChat` continuation calls.

**Comment by kouloumos** [comment]
I [force-pushed](https://github.com/schemalabz/opencouncil-tasks/compare/ecb74e4573f3e311b0d0b17950c8989132f12410..a79a6e63dceaed0098cc900d71c8bc980bbf28bd) to address bot review comment(s) ([1](https://github.com/schemalabz/opencouncil-tasks/pull/33#discussion_r3043659202), [2](https://github.com/schemalabz/opencouncil-tasks/pull/33#discussion_r3095457695), [3](https://github.com/schemalabz/opencouncil-tasks/pull/33#discussion_r3095699752), [4](https://github.com/schemalabz/opencouncil-tasks/pull/33#discussion_r3095726117), [5](https://github.com/schemalabz/opencouncil-tasks/pull/33#discussion_r3095726123)):

- Fix off-by-one in tail-page extraction guard (`>=` instead of `>`)
- Catch empty-string `decisionNumber` in validation warnings
- Add `name` field validation on LLM match response
- Forward `parseJson` parameter in `aiChat` continuation calls
- Pass full people list to LLM fallback and deduplicate vote details by personId

File: src/tasks/utils/decisionPdfExtraction.ts [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+            return { result, usage: totalUsage, fromCache: false };
+        }
+
+        console.log(`  Incomplete extraction — decision content not found in first ${actualPages} pages, retrying with more...`);
+        pagesToSend += PAGE_INCREMENT;
+    }
+
+    // Front pages exhausted — try the last TAIL_PAGES pages
+    const tailStart = Math.max(0, totalPages - TAIL_PAGES);
+    if (tailStart > MAX_FRONT_PAGES) {
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/tasks/utils/decisionPdfExtraction.ts]: Fixed in [force-push](https://github.com/schemalabz/opencouncil-tasks/compare/ecb74e4573f3e311b0d0b17950c8989132f12410..a79a6e63dceaed0098cc900d71c8bc980bbf28bd) — changed `>` to `>=`

File: src/tasks/utils/decisionValidation.ts [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+        });
+    }
+
+    if (raw.decisionNumber == null) {
+        warnings.push({
+            code: 'MISSING_DECISION_NUMBER',
+            severity: 'info',
+            message: 'Decision number (Αριθμός Απόφασης) not found in document',
+        });
+    }
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/tasks/utils/decisionValidation.ts:63]: Fixed in [force-push](https://github.com/schemalabz/opencouncil-tasks/compare/ecb74e4573f3e311b0d0b17950c8989132f12410..a79a6e63dceaed0098cc900d71c8bc980bbf28bd) — changed `== null` to `!raw.decisionNumber`

File: src/tasks/utils/decisionPdfExtraction.ts [hunk 3]
```diff
@@ -0,0 +1,587 @@
...
+    const usedIds = new Set<string>();
+
+    for (const entry of result) {
+        if (entry.personId && !usedIds.has(entry.personId)) {
+            matched.push({ name: entry.name, personId: entry.personId });
+            usedIds.add(entry.personId);
+        } else {
+            stillUnmatched.push(entry.name);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/tasks/utils/decisionPdfExtraction.ts:404]: Fixed in [force-push](https://github.com/schemalabz/opencouncil-tasks/compare/ecb74e4573f3e311b0d0b17950c8989132f12410..a79a6e63dceaed0098cc900d71c8bc980bbf28bd) — added guard to skip entries with missing/invalid `name`

File: src/lib/ai.ts [hunk 4]
```diff
@@ -206,6 +215,7 @@ export async function aiChat<T>({ model, systemPrompt, userPrompt, prefillSystem
                 userPrompt,
                 prefillSystemResponse: ((prefillSystemResponse || '') + responseText).trim(),
                 prependToResponse: ((prependToResponse || '') + responseText).trim(),
+                maxTokens: maxTokensParam,
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/lib/ai.ts:219]: Fixed in [force-push](https://github.com/schemalabz/opencouncil-tasks/compare/ecb74e4573f3e311b0d0b17950c8989132f12410..a79a6e63dceaed0098cc900d71c8bc980bbf28bd) — forwarded `parseJson` in the recursive `aiChat` call

File: src/tasks/utils/extractionPipeline.ts [hunk 5]
```diff
@@ -0,0 +1,237 @@
...
+
+    console.log(`\n--- Meeting-level matching ---`);
+    console.log(`  Unique names: ${allRawNames.size}`);
+    console.log(`  Token-sort matched: ${nameToPersonId.size}`);
+    console.log(`  Remaining for LLM: ${step1Unmatched.length}`);
+
+    // Step 2: LLM fallback for remaining unmatched
+    if (step1Unmatched.length > 0) {
+        const matchedIds = new Set(nameToPersonId.values());
+        const availablePeople = people.filter(p => !matchedIds.has(p.id));
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/tasks/utils/extractionPipeline.ts]: Fixed in [force-push](https://github.com/schemalabz/opencouncil-tasks/compare/ecb74e4573f3e311b0d0b17950c8989132f12410..a79a6e63dceaed0098cc900d71c8bc980bbf28bd) — removed the `availablePeople` filter so the LLM can match abbreviated vote-detail names to already-matched people, and added vote detail deduplication by personId to handle the resulting name variants

File: src/tasks/utils/decisionPdfExtraction.ts [hunk 6]
```diff
@@ -0,0 +1,588 @@
...
+            return { result, usage: totalUsage, fromCache: false };
+        }
+
+        console.log(`  Incomplete extraction — decision content not found in first ${actualPages} pages, retrying with more...`);
+        pagesToSend += PAGE_INCREMENT;
+    }
+
+    // Front pages exhausted — try the last TAIL_PAGES pages
+    const tailStart = Math.max(0, totalPages - TAIL_PAGES);
+    if (tailStart >= MAX_FRONT_PAGES) {
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/tasks/utils/decisionPdfExtraction.ts:544]: False positive — for 16–19 page PDFs, `tailStart` (11–14) falls within the range already covered by front-page extraction (0–14). The guard correctly prevents re-sending overlapping pages. If the decision content wasn't found in the first 15 pages and the tail overlaps, `incomplete: true` is the expected outcome.

File: src/tasks/utils/extractionPipeline.ts [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+
+    console.log(`\n--- Meeting-level matching ---`);
+    console.log(`  Unique names: ${allRawNames.size}`);
+    console.log(`  Token-sort matched: ${nameToPersonId.size}`);
+    console.log(`  Remaining for LLM: ${step1Unmatched.length}`);
+
+    // Step 2: LLM fallback for remaining unmatched
+    if (step1Unmatched.length > 0) {
+        try {
+            const llmResult = await llmMatchMembers(step1Unmatched, people);
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/tasks/utils/extractionPipeline.ts:140]: False positive — this is an intentional change from the [previous force-push](https://github.com/schemalabz/opencouncil-tasks/compare/ecb74e4573f3e311b0d0b17950c8989132f12410..a79a6e63dceaed0098cc900d71c8bc980bbf28bd). Passing the full people list allows the LLM to match abbreviated vote-detail names (e.g. `Παπαδόπουλος Ι.`) to people already matched by token-sort under their formal name. Duplicate personIds in `nameToPersonId` are expected (two name variants → same person). Vote detail deduplication via `seenVoterIds` handles the downstream concern.

File: src/types.ts [hunk 8]
```diff
@@ -406,19 +406,56 @@ export interface GenerateVoiceprintResult {
...
- * Task: Poll Decisions (Diavgeia)
+ * Extract Decisions (PDF → structured data)
+ */
+
+/** Per-decision warning from the extraction pipeline. See DecisionWarningCode in decisionValidation.ts for the full list of codes. */
+export interface DecisionWarning {
+    code: string;
+    severity: 'info' | 'warning' | 'error';
+    message: string;
+}
```
[the remark this answers is not in the document]
Comment (kouloumos) [reply src/types.ts:417]: Valid observation but outside the scope of this PR.
